### PR TITLE
added module ORM/Hibernate

### PR DESCRIPTION
**Pull request by Knarik8**

### CODE REVIEW

File: src/main/java/epam/gym/config/JPAConfig.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import org.springframework.transaction.annotation.EnableTransactionManagement;
+
+import javax.sql.DataSource;
+import java.util.Properties;
+
+
+@Configuration
+@ComponentScan("epam.gym")
+@PropertySource("classpath:application.properties")
+@EnableTransactionManagement
```
Comment (sergeypil) [inline src/main/java/epam/gym/config/JPAConfig.java]: I believe we can remove these annotations for one of config classes not to dublicate
@ComponentScan("epam.gym")
@PropertySource("classpath:application.properties")
@EnableTransactionManagement

File: src/main/java/epam/gym/dao/hibernateImpl/TraineeDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    @Override
+    @Transactional
+    public Trainee update(Trainee trainee) {
+        return entityManager.merge(trainee);
+    }
+
+    @Override
+    @Transactional
```
Comment (sergeypil) [inline src/main/java/epam/gym/dao/hibernateImpl/TraineeDaoImpl.java]: In most cases, @Transactional is not required for read operations

File: src/main/java/epam/gym/dao/hibernateImpl/TraineeDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+    @Override
+    public Set<String> getExistingUsernames() {
+        List<String> usernames = entityManager.createQuery("SELECT username FROM Trainee", String.class).getResultList();
+        return new HashSet<>(usernames);
+    }
+
+    @Override
+    @Transactional
+    public Optional<Trainee> selectTraineeByUsername(String username) {
```
Comment (sergeypil) [inline src/main/java/epam/gym/dao/hibernateImpl/TraineeDaoImpl.java]: I'd follow the single format of method names `findTraineeByUsername`

File: src/main/java/epam/gym/dao/hibernateImpl/TraineeDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        try {
+            Trainee trainee = entityManager.createQuery("SELECT t FROM Trainee t LEFT JOIN FETCH t.trainings WHERE " +
+                            "t.username = :username", Trainee.class)
+                    .setParameter("username", username)
+                    .getSingleResult();
+            return Optional.of(trainee);
+        } catch (Exception e) {
+            return Optional.empty();
+        }
```
Comment (sergeypil) [inline src/main/java/epam/gym/dao/hibernateImpl/TraineeDaoImpl.java:72]: Probably we do need catch exception here.
`        } catch (Exception e) {
            return Optional.empty();
        }`
But if we catch, we should log it or rethrow

File: src/main/java/epam/gym/dao/hibernateImpl/TraineeDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+    @PersistenceContext
+    private EntityManager entityManager;
+
+    @Override
+    @Transactional
+    public Trainee create(Trainee trainee) {
+        entityManager.persist(trainee);
+        return trainee;
+    }
```
Comment (sergeypil) [inline src/main/java/epam/gym/dao/hibernateImpl/TraineeDaoImpl.java:33]: Sometimes it is good practice to add @NonNull annotations for such params
   ```
 @Override
    @Transactional
    public Trainee create(@NonNull Trainee trainee) {
        entityManager.persist(trainee);
        return trainee;
    }
```

File: src/main/java/epam/gym/service/TrainerService.java [hunk 6]
```diff
@@ -1,10 +1,31 @@
 package epam.gym.service;
 
 import epam.gym.entity.Trainer;
+import epam.gym.entity.Training;
+
+import javax.naming.AuthenticationException;
```
Comment (sergeypil) [inline src/main/java/epam/gym/service/TrainerService.java]: I think we can create our own exception AuthenticationException, which extends RunTimeException. And not to add throws AuthenticationException everywhere after method

File: src/main/java/epam/gym/service/hibernateImpl/TraineeServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+    @Override
+    public void deactivateTrainee(Long id, String username, String password) throws AuthenticationException {
+        if (authenticationService.authenticate(id, username, password)) {
+            Optional<Trainee> traineeOpt = findById(id, username, password);
+            if (traineeOpt.isPresent()) {
+                traineeDao.setActiveStatus(id, false);
+                logger.info("Deactivated trainee with ID: {}", id);
+            } else {
+                logger.error("Trainee not found with ID: {}", id);
+                throw new IllegalArgumentException("Trainee not found");
```
Comment (sergeypil) [inline src/main/java/epam/gym/service/hibernateImpl/TraineeServiceImpl.java:157]: Create own exception, EntityNotFoundException for example. IllegalArgumentException is used when bad arguments in method

File: src/main/java/epam/gym/dao/hibernateImpl/TraineeDaoImpl.java (hunk 3 above)
Comment (Knarik8) [reply src/main/java/epam/gym/dao/hibernateImpl/TraineeDaoImpl.java]: fixed

File: src/main/java/epam/gym/dao/hibernateImpl/TraineeDaoImpl.java (hunk 2 above)
Comment (Knarik8) [reply src/main/java/epam/gym/dao/hibernateImpl/TraineeDaoImpl.java]: deleted from read methods

File: src/main/java/epam/gym/dao/hibernateImpl/TraineeDaoImpl.java (hunk 4 above)
Comment (Knarik8) [reply src/main/java/epam/gym/dao/hibernateImpl/TraineeDaoImpl.java:72]: added logging

File: src/main/java/epam/gym/dao/hibernateImpl/TraineeDaoImpl.java (hunk 5 above)
Comment (Knarik8) [reply src/main/java/epam/gym/dao/hibernateImpl/TraineeDaoImpl.java:33]: added for some methods

File: src/main/java/epam/gym/service/TrainerService.java (hunk 6 above)
Comment (Knarik8) [reply src/main/java/epam/gym/service/TrainerService.java]: added own exception AuthenticationException, deleted throws AuthenticationException after methods

**Review by sergeypil — APPROVED** [review]

**Pull request merged**